packages/story-editor/src/app/taxonomy/taxonomyProvider.tsx

```diff
@@ -0,0 +1,285 @@
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from '@googleforcreators/react';
+import type { PropsWithChildren } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import type {
+  createTermProps,
+  setTermsProps,
+  TaxonomyRestError,
+  addTermToSelectionProps,
+  addSearchResultsToCacheProps,
+  removeTermsProps,
+  UpdateStoryProps,
+} from '../../types';
+import cleanForSlug from '../../utils/cleanForSlug';
+import { useAPI } from '../api';
+import { useStory } from '../story';
+import Context from './context';
+
+function TaxonomyProvider(props: PropsWithChildren<unknown>) {
+  // Should grab categories on mount
+  const [shouldRefetchCategories, setShouldRefetchCategories] =
+    useState<boolean>(true);
+  const { updateStory, terms } = useStory((state) => ({
+    updateStory: state.actions.updateStory,
+    terms: state.state.story?.terms || [],
+  }));
+  const [hasTaxonomies, setHasTaxonomies] = useState<boolean>(false);
+  const [taxonomies, setTaxonomies] = useState<Taxonomy[]>([]);
+  const [termCache, setTermCache] = useState<Term[]>(terms);
+
+  const {
+    actions: { getTaxonomyTerm, createTaxonomyTerm, getTaxonomies },
+  } = useAPI();
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    if (hasTaxonomies || !getTaxonomies) {
+      return;
+    }
+
+    void (async () => {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(result);
+      } catch (e) {
+        if (e instanceof Error) {
+          // eslint-disable-next-line no-console -- Log error
+          console.error(e.message);
+        }
+      } finally {
+        setHasTaxonomies(true);
+      }
+    })();
+  }, [hasTaxonomies, getTaxonomies]);
+
+  const setTerms = useCallback(
+    ({ newTerms }: setTermsProps) => {
+      if (updateStory) {
+        const properties = (story: Story) => {
+          const currentTerms = story?.terms || [];
+          const newAssignedTerms = [...new Set([...currentTerms, ...newTerms])];
+          return {
+            ...story,
+            terms: newAssignedTerms,
+          };
+        };
+        updateStory({ properties } as UpdateStoryProps);
+      }
+    },
+    [updateStory]
+  );
+
+  const removeTerms = useCallback(
+    ({ removeTerms }: removeTermsProps) => {
+      if (updateStory) {
+        const removeTermsID = removeTerms.map(({ id }) => id);
+        const properties = (story: Story) => {
+          const currentTerms = story?.terms || [];
+          const newAssignedTerms = currentTerms.filter(
+            ({ id }) => !removeTermsID.includes(id)
+          );
+          return {
+            ...story,
+            terms: newAssignedTerms,
+          };
+        };
+        updateStory({ properties } as UpdateStoryProps);
+      }
+    },
+    [updateStory]
+  );
+
+  const addTermToSelection = useCallback(
+    ({ newTerms }: addTermToSelectionProps) => setTerms({ newTerms }),
+    [setTerms]
+  );
```

Comment:
That's an odd function signature, why not just

```suggestion
  const addTermToSelection = useCallback(
    (newTerms: Term[]) => setTerms({ newTerms }),
    [setTerms]
  );
```

Ditto with `setTerms` above